langchain4j-core/src/main/java/dev/langchain4j/rag/content/retriever/WebSearchContentRetriever.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Constructs a new WebSearchContentRetriever with the specified web search engine, embedding model, and maximum results.
+     * @param webSearchEngine The web search engine to use for retrieving search results.
+     * @param embeddingModel The embedding model to use for generating embeddings.
+     * @param maxResults The maximum number of relevant results to retrieve.
+     */
+    public WebSearchContentRetriever(WebSearchEngine webSearchEngine, EmbeddingModel embeddingModel, Integer maxResults) {
+        this(webSearchEngine, embeddingModel, null, maxResults);
+    }
+
+    /**
+     * Constructs a new WebSearchContentRetriever with the specified web search engine, embedding model, and document splitter.
+     * @param webSearchEngine The web search engine to use for retrieving search results.
+     * @param embeddingModel The embedding model to use for generating embeddings and finding the most relevant 3 web search results.
+     * @param documentSplitter The document splitter to use for splitting search results into text segments.
+     */
+    public WebSearchContentRetriever(WebSearchEngine webSearchEngine,
+                                     EmbeddingModel embeddingModel,
+                                     DocumentSplitter documentSplitter) {
+        this(webSearchEngine, embeddingModel, documentSplitter, DEFAULT_MAX_RESULTS);
+    }
+
+    /**
+     * Constructs a new WebSearchContentRetriever with the specified web search engine, embedding model, document splitter, and maximum results.
+     * @param webSearchEngine The web search engine to use for retrieving search results.
+     * @param embeddingModel The embedding model to use for generating embeddings and finding the relevant web search results.
+     * @param documentSplitter The document splitter to use for splitting search results into text segments.
+     * @param maxResults The maximum number of relevant results to retrieve.
+     */
+    @Builder
+    public WebSearchContentRetriever(WebSearchEngine webSearchEngine,
+                                     EmbeddingModel embeddingModel,
+                                     DocumentSplitter documentSplitter,
+                                     Integer maxResults) {
+        this.webSearchEngine = ensureNotNull(webSearchEngine, "webSearchEngine");
+        this.embeddingModel = embeddingModel;
+        this.documentSplitter = documentSplitter;
+        this.embeddingStore = new InMemoryEmbeddingStore<>();
+        this.maxResults = ensureGreaterThanZero(getOrDefault(maxResults, DEFAULT_MAX_RESULTS), "maxResults");
+    }
+
+    @Override
+    public List<Content> retrieve(Query query) {
+        WebSearchResults webSearchResults = webSearchEngine.search(query.text());
+        List<TextSegment> textSegments = webSearchResults.toTextSegments();
+        if (documentSplitter != null) {
+            textSegments = documentSplitter.splitAll(webSearchResults.toDocuments());
```

Comment:
On a second thought, it actually seems to be a very good idea to give an option to just split pages without further search/filter, so that in the next step, re-ranking (and filtering) can be done with a much better model!

I guess this can be a third constructor: `(WebSearchEngine, DocumentSplitter)`

Reply:
yeah exactly and with a much better `embeddingStore` to search/filter using a better technique where user be able to set a `minScore` to retrieve more accurate `textSegments` in a next step.